pkg/base/helm.go

```diff
@@ -243,57 +254,168 @@ func removeCommonPrefix(baseFiles []BaseFile) []BaseFile {
-				Path:    "Chart.lock",
-				Content: upstreamFile.Content,
-			})
-		} else if base.Path != "" && strings.Contains(upstreamFile.Path, path.Join(base.Path, "Chart.lock")) {
-			base.AdditionalFiles = append(base.AdditionalFiles, BaseFile{
-				Path:    "Chart.lock",
-				Content: upstreamFile.Content,
-			})
+// creates a map of the base paths and their cooresponding upstream paths
+func getBaseToUpstreamPathMap(upstreamFileMap map[string][]byte, upstreamToBasePathsMap map[string][]string) map[string]string {
+	baseToUpstreamPathMap := make(map[string]string)
+	for upstreamFilePath, basePaths := range upstreamToBasePathsMap {
+		for _, basePath := range basePaths {
+			baseToUpstreamPathMap[basePath] = upstreamFilePath
 		}
 	}
+	return baseToUpstreamPathMap
+}
+
+// adds the Chart.yaml and Chart.lock files from the upstream to the base charts that were rendered
+func helmChartBaseAppendAdditionalFiles(base Base, fullBasePath string, upstreamFiles map[string][]byte, baseToUpstreamPathMap map[string]string) Base {
+	if upstreamPath, ok := baseToUpstreamPathMap[fullBasePath]; ok {
+		additionalFiles := []string{"Chart.yaml", "Chart.lock"}
+		for _, additionalFile := range additionalFiles {
+			additionalFilePath := path.Join(upstreamPath, additionalFile)
+			if content, ok := upstreamFiles[additionalFilePath]; ok {
+				base.AdditionalFiles = append(base.AdditionalFiles, BaseFile{
+					Path:    additionalFile,
+					Content: content,
+				})
+			}
+		}
+	} else {
+		logger.Errorf("failed to find upstream path for base path %s", fullBasePath)
+	}
 
 	var nextBases []Base
-	for _, base := range base.Bases {
-		base = helmChartBaseAppendAdditionalFiles(base, u)
-		nextBases = append(nextBases, base)
+	for _, nextBase := range base.Bases {
+		nextBase = helmChartBaseAppendAdditionalFiles(nextBase, path.Join(base.Path, nextBase.Path), upstreamFiles, baseToUpstreamPathMap)
+		nextBases = append(nextBases, nextBase)
 	}
 	base.Bases = nextBases
 
 	return base
 }
 
-// look for any sub-chart dependencies that are missing from base and add their Chart.yaml
-func helmChartBaseAppendMissingDependencies(base Base, u *upstreamtypes.Upstream) Base {
-	basePaths := getAllBasePaths("", base)
-	basePathMap := map[string]bool{}
-	for _, basePath := range basePaths {
-		basePathMap[basePath] = true
+func helmChartUpstreamPathToBasePaths(upstreamPath string, upstreamFileMap map[string][]byte) ([]string, error) {
+	charts := pathToCharts(upstreamPath)
+	basePathParts := make([][]string, len(charts))
+	basePathParts[0] = []string{""}
+	// iterate over the subcharts and find any that are aliased
+	for i := 1; i < len(charts); i++ {
+		parts := []string{}
+		chart := charts[i]
+		parts = append(parts, chart)
+		parentChart := charts[i-1]
+
+		// build the full parent chart path and check if the chart is aliased by its parent
+		parentChartYaml := filepath.Join(parentChart, "Chart.yaml")
+		for j := 0; j < i-1; j++ {
```

Comment:
i think this would break if you have a sub-sub-sub-chart like:
```
charts/mariadb/charts/common/charts/anothercharthere
```
i'd suggest tracking the full path in a variable outside of the loop

Reply:
Done. This also got me thinking that the whole `flattenBasePathParts` workflow is not necessary since I am already iterating through all of the possible base branches for the upstream path here.